doc/cprover-manual/contracts-loops.md

```diff
@@ -0,0 +1,118 @@
+# Loop Contracts
+
+CBMC offers support for loop contracts, which includes three basic clauses:
+- _invariant_ clause for establishing safety properties
+- _decreases_ clause for establishing termination, and
+- _assigns_ clause for declaring the subset of variables that is modifiable in the loop.
+
+These clauses formally describe an abstraction of a loop for the purpose of a proof.
+CBMC also provides a series of built-in constructs
+to aid writing loop contracts (e.g., _history variables_ and _quantifiers_).
+
+## Overview
+
+Consider an implementation of the [binary search algorithm] below.
+
+```c
+#include <assert.h>
+
+int binary_search(int val, int buf[], int size)
+{
+  if(size < 0) return -1;
+
+  long lb = 0, ub = size - 1;
+  long mid = (lb + ub) / 2;
+
+  while(lb <= ub)
+  {
+     if(buf[mid] == val) break;
+     if(buf[mid] < val)
+       lb = mid + 1;
+     else
+       ub = mid - 1;
+     mid = (lb + ub) / 2;
+  }
+  assert(lb > ub || (0 <= mid && mid < size));
+  return lb > ub ? -1 : mid;
+}
+```
+
+The function stores a lower bound `lb` and an upper bound `ub`
+initialized to the bounds on the buffer `buf`, i.e., to `0` and `size-1` respectively.
+In each iteration, the midpoint `mid` is compared against the target value `val`
+and in case of a mismatch either the lower half or the upper half of the buffer is searched recursively.
+A developer might be interested in verifying two high-level properties on the loop on all possible buffers `buf` and values `val`:
+1. an out-of-bound access should never occur (at `buf[mid]` lookup)
+2. the loop must eventually always terminate
+
+To prove the first (memory-safety) property,
+we may declare _loop invariants_ that must be preserved across all loop iterations.
+In this case, two invariant clauses would together imply that `buf[mid]` lookup is always safe.
+The first invariant clause would establish that the bounds (`lb` and `ub`) are always valid:
+```c
+__CPROVER_loop_invariant(0 <= lb && lb - 1 <= ub && ub < size)
+```
+Note that in the second conjunct, `lb - 1 == ub` case is possible when the value `val` is not found in the buffer `buf`.
+The second invariant clause would establish that the midpoint `mid` is always a valid index.
+In this particular case we can in fact establish a stronger invariant,
+that `mid` is indeed always the midpoint of `lb` and `ub` in every iteration:
+```c
+__CPROVER_loop_invariant(mid == (lb + ub) / 2)
```

Comment:
Thanks; you are right about the weaker invariant. The performance is quite good already though -- I ran the same commands I specified in this file and CBMC generates an unbounded proof in fraction of a second. 
I'll keep the current abstraction since it's stronger and seems easier to connect it to the implementation.